sklearn/decomposition/_lda.py

```diff
@@ -539,8 +547,18 @@ def _check_non_neg_array(self, X, reset_n_features, whom):
         X :  array-like or sparse matrix
 
         """
-        X = self._validate_data(X, reset=reset_n_features, accept_sparse="csr")
+        X = self._validate_data(
+            X,
+            reset=reset_n_features,
+            accept_sparse="csr",
+            dtype=[np.float64, np.float32],
+        )
         check_non_negative(X, whom)
+
+        # Ensure input data has same dtype with trained components_
+        if hasattr(self, "components_") and self.components_ is not None:
+            X = X.astype(self.components_.dtype)
```

Comment:
Is this really necessary? Why not keep the up-cast if the users decides to use `np.float64` at prediction time (even if the model was fit with `np.float32` data?

If this condition is really necessary, we should avoid a introducing a new memory copy when not necessary:

```suggestion
            X = X.astype(self.components_.dtype, copy=False)
```

Reply:
The code is introduced to ensure only one dtype used in interior implementation `mean_change` and `_dirichlet_expectation_1d` in `_online_lda_fast.pyx` require two arrays which have same dtype. 

But for now I have changed `_online_lda_fast.pyx` to accept different two dtypes as input. As a result there is no need to align dtype, so I have remove the code.

Reply:
@ogrisel Could you review again if needed?